marjapussi/agent.py

```diff
@@ -65,21 +71,34 @@ def observe_action(self, action):
                 self.state['possible_cards'] = self._possible_cards_after_trick(self.state['possible_cards'], self.state['current_trick'])
                 self.state['current_trick'] = []
 
+            if len(self.state['all_tricks']) == 0 and len(self.state['current_trick']) == 0:
+                if val.split('-')[1] != 'A':
+                    for i in COLORS:
+                        self.state['possible_cards'][player_name].discard(f"{i}-A")
+                elif val.split('-')[0] != 'g':
+                    for i in VALUES:
+                        self.state['possible_cards'][player_name].discard(f"g-{i}")
+
         if phase == 'PASS' or phase == "PBCK":
             self.state['secure_cards'][player_name].discard(val)
             self.state['possible_cards'][player_name].discard(val)
-            self.state['secure_cards'][self.all_players[(player_num + 2)%4]].add(val)
-            self.state['possible_cards'][self.all_players[(player_num + 2)%4]].add(val)
+            self.state['secure_cards'][self.all_players[partner_num]].add(val)
```

Comment:
Souldn't this not be partner_num? The idea is to update the knowledge about the enemies.